lib/fluent/plugin/out_forward.rb

```diff
@@ -751,7 +772,179 @@ def resolved_host
+        # those with weight 0 are placed at the beginning of the list.
+        shuffle_start = nil
+        srv_list.sort_by!(&:weight)
+        srv_list.each_index do |index|
+          if srv_list[index].weight != 0
+            if shuffle_start.nil?
+              shuffle_start = index
+            end
+            if !shuffle_start.nil? && srv_list[index].weight != 0
+              shuffle_end = index
+              target = srv_list.dup
+              srv_list[shuffle_start..shuffle_end] = target[shuffle_start..shuffle_end].shuffle!
+            end
+          end
+        end
+
+        srv_list
+      end
+
+      # @param [SRV] srv_list
+      # @return [SRV]
+      def shuffle_by_weight_srv_servers(srv_list)
+        return [] if srv_list.nil? || srv_list.empty?
+
+        # Compute the sum of the weights of those RRs, and with each RR
+        # associate the running sum in the selected order
+        sum = srv_list.inject(0) { |sum, srv| sum + srv.weight}
+
+        srv_list = random_order_with_weight(srv_list)
+        ret = []
+        # Then choose a uniform random number between 0 and the sum computed (inclusive),
+        # and select the RR whose running sum value is the first in the selected order which is greater than or equal to the random number selected.
+        until sum <= 0 || srv_list.empty?
+          selector = Integer(rand(sum))
+          running_sum = 0
+
+          srv_list.each_index do |index|
+            running_sum += srv_list[index].weight
+            next unless running_sum > selector
+            ret.push srv_list[index]
+            sum -= srv_list[index].weight
+            srv_list.delete_at index
+            break
+          end
+        end
+        ret
+      end
+
+      def switch_original_host!
+        @srv_host_mutex.synchronize do
+          if @using_srv
+            @host = @original_host
+            @port = @original_port
+            @using_srv = false
+          end
+        end
+      end
+
+      def resolve_srv!
+        if @using_srv
+          switch_original_host!
+        end
+
+        host = "_#{@srv_service_name}._#{@srv_service_protocol}.#{@host}"
+        @log.info 'srv try resolve hostname', host: host
+        resp = resolve_srv(host)
+
+        if resp.empty?
+          # empty response is tried original host.
+          @log.warn 'srv record empty response', host: host
+          if @using_srv
+            switch_original_host!
+          end
+          return
+        end
+
+        resp = srv_list_sort_priority_weight(resp)
+        available = resp.find(&:available?)
+
+        if available.nil?
+          # unavailable is tried original host.
+          if @using_srv
+            switch_original_host!
+          end
+          @log.warn 'srv record does not available node. try using A record', host: host
+          return
+        end
+
+        # swap config host to srv response host.
+        @srv_host_mutex.synchronize do
```

Comment:
We need to lock other places where node uses `@host` and `@port`(e.g. https://github.com/fluent/fluentd/pull/2496/files#diff-40c0e8c88efdc9a297a9fc63bc2bc69fR948).
This code can be a race condition.

1. t1 tries to do [this line](https://github.com/fluent/fluentd/pull/2496/files#diff-40c0e8c88efdc9a297a9fc63bc2bc69fR948) and evaluates arguments `@host`
2. context switch happens to t2
3. t2 does L930~L938,(and set `@host` and `@port` to another value)
4. context switch happens to t1
5. `@port` is invalid for evaluating `@host` at step 1 since `@port` has changed by t2